src/main/java/com/zendesk/maxwell/MaxwellParser.java

```diff
@@ -133,71 +126,122 @@ private MaxwellAbstractRowsEvent processRowsEvent(AbstractRowEvent e) throws Sch
+				break;
 			case MySQLConstants.UPDATE_ROWS_EVENT:
+				ew = new MaxwellUpdateRowsEvent((UpdateRowsEvent) e, table, filter);
+				break;
 			case MySQLConstants.UPDATE_ROWS_EVENT_V2:
+				ew = new MaxwellUpdateRowsEvent((UpdateRowsEventV2) e, table, filter);
+				break;
 			case MySQLConstants.DELETE_ROWS_EVENT:
+				ew = new MaxwellDeleteRowsEvent((DeleteRowsEvent) e, table, filter);
+				break;
 			case MySQLConstants.DELETE_ROWS_EVENT_V2:
-				rowEventsProcessed++;
-				event = processRowsEvent((AbstractRowEvent) v4Event);
-				if ( event.matchesFilter() )
-					return event;
+				ew = new MaxwellDeleteRowsEvent((DeleteRowsEventV2) e, table, filter);
 				break;
-			case MySQLConstants.TABLE_MAP_EVENT:
-				if ( stopAtNextTableMap)
-					return null;
+			default:
+				return null;
+		}
+		return ew;
+	}
 
-				tableCache.processEvent(this.schema, (TableMapEvent) v4Event);
-				break;
-			case MySQLConstants.QUERY_EVENT:
-				processQueryEvent((QueryEvent) v4Event);
+	private LinkedList<MaxwellAbstractRowsEvent> getTransactionEvents() throws Exception {
+		BinlogEventV4 v4Event;
+		MaxwellAbstractRowsEvent event;
+
+		LinkedList<MaxwellAbstractRowsEvent> list = new LinkedList<>();
+
+		// currently to satisfy the test interface, the contract is to return null
+		// if the queue is empty.  should probably just replace this with an optional timeout...
+
+		while ( true ) {
+			v4Event = queue.poll(100, TimeUnit.MILLISECONDS);
+			if (v4Event == null)
+				continue;
+
+			switch(v4Event.getHeader().getEventType()) {
+				case MySQLConstants.WRITE_ROWS_EVENT:
+				case MySQLConstants.WRITE_ROWS_EVENT_V2:
+				case MySQLConstants.UPDATE_ROWS_EVENT:
+				case MySQLConstants.UPDATE_ROWS_EVENT_V2:
+				case MySQLConstants.DELETE_ROWS_EVENT:
+				case MySQLConstants.DELETE_ROWS_EVENT_V2:
+					rowEventsProcessed++;
+					event = processRowsEvent((AbstractRowEvent) v4Event);
+
+					if ( event.matchesFilter() )
+						list.add(event);
+					break;
+				case MySQLConstants.TABLE_MAP_EVENT:
+					tableCache.processEvent(this.schema, (TableMapEvent) v4Event);
+					break;
+				case MySQLConstants.QUERY_EVENT:
+					QueryEvent qe = (QueryEvent) v4Event;
+					// TODO: need to handle this case.
+					// The MySQL guys say some storage engines will output a "COMMIT" QUERY_EVENT at the
+					// end of the stream.
+					throw new RuntimeException("Unhandled QueryEvent: " + qe);
+				case MySQLConstants.XID_EVENT:
+					XidEvent xe = (XidEvent) v4Event;
+					for ( MaxwellAbstractRowsEvent e : list )
+						e.setXid(xe.getXid());
+
+					if ( !list.isEmpty() )
+						list.getLast().setCommit(true);
+
+					return list;
 			}
 		}
 	}
 
+	private LinkedList<MaxwellAbstractRowsEvent> txBuffer;
+
 	public MaxwellAbstractRowsEvent getEvent() throws Exception {
-		return getEvent(false);
+		BinlogEventV4 v4Event;
+		MaxwellAbstractRowsEvent event;
+
+		while ( true ) {
+			if ( txBuffer != null && !txBuffer.isEmpty() ) {
+				return txBuffer.removeFirst();
```

Comment:
yeah, I wanted to keep the interface with `getEvent`, even though we really spool up a whole transaction's worth.  I may change this to simply return a whole transaction's worth of rows -- we really only want to set a binlog checkpoint after a transaction -- but for now it's like this.

i also sort of wonder if we'd end up overflowing the heap and crashing on really really big 100mb transactions.  Maybe I should read to the XID_EVENT and then backtrack? hmmmmm